grype/matcher/rpmdb/matcher_test.go

```diff
@@ -7,74 +7,129 @@ import (
-
-	foundCVEs := internal.NewStringSet()
-
-	for _, a := range actual {
-		foundCVEs.Add(a.Vulnerability.ID)
-
-		assert.Equal(t, match.ExactIndirectMatch, a.Type, "indirect match not indicated")
-		assert.Equal(t, p.Name, a.Package.Name, "failed to capture original package name")
-		for _, detail := range a.MatchDetails {
-			assert.Equal(t, matcher.Type(), detail.Matcher, "failed to capture matcher type")
-		}
-	}
-
-	for _, id := range []string{"CVE-2014-fake-2", "CVE-2013-fake-3"} {
-		if !foundCVEs.Contains(id) {
-			t.Errorf("missing discovered CVE: %s", id)
-		}
-	}
-	if t.Failed() {
-		t.Logf("discovered CVES: %+v", foundCVEs)
-	}
-
-}
-
-func TestMatcherDpkg_matchBySourceIndirection_ignoreSource(t *testing.T) {
-	matcher := Matcher{}
-	p := pkg.Package{
-		Name:    "neutron",
-		Version: "7.1.3-6",
-		Type:    syftPkg.RpmPkg,
-		Metadata: pkg.RpmdbMetadata{
-			// this ends up being the same matches as the original package, thus should be ignored
-			SourceRpm: "neutron-7.1.3-6.el8.src.rpm",
+		{
+			name: "Rpmdb Match matches by direct and ignores the source rpm when the package names are the same",
+			p: pkg.Package{
+				Name:    "neutron",
+				Version: "7.1.3-6",
+				Type:    syftPkg.RpmPkg,
+				Metadata: pkg.RpmdbMetadata{
+					SourceRpm: "neutron-7.1.3-6.el8.src.rpm",
+				},
+			},
+			setup: func() (vulnerability.Provider, distro.Distro, Matcher) {
+				matcher := Matcher{}
+				d, err := distro.NewDistro(distro.CentOS, "8", "")
+				if err != nil {
+					t.Fatal("could not create distro: ", err)
+				}
+
+				store := newMockProvider("neutron", "neutron-devel")
+
+				return store, d, matcher
+			},
+			expectedMatches: map[string]match.Type{
+				"CVE-2014-fake-1": match.ExactDirectMatch,
+			},
+		},
+		{ // Regression against https://github.com/anchore/grype/issues/376
+			name: "Rpmdb Match matches by direct and by source indirection when the SourceRpm version is desynced from package version",
+			p: pkg.Package{
+				Name:    "neutron-libs",
+				Version: "7.1.3-6",
+				Type:    syftPkg.RpmPkg,
+				Metadata: pkg.RpmdbMetadata{
+					SourceRpm: "neutron-17.16.3-229.el8.src.rpm",
+				},
+			},
+			setup: func() (vulnerability.Provider, distro.Distro, Matcher) {
+				matcher := Matcher{}
+				d, err := distro.NewDistro(distro.CentOS, "8", "")
+				if err != nil {
+					t.Fatal("could not create distro: ", err)
+				}
+
+				store := newMockProvider("neutron-libs", "neutron")
+
+				return store, d, matcher
+			},
+			expectedMatches: map[string]match.Type{
+				"CVE-2014-fake-1": match.ExactDirectMatch,
+			},
+			wantErr: false,
 		},
 	}
 
-	d, err := distro.NewDistro(distro.CentOS, "8", "")
-	if err != nil {
-		t.Fatal("could not create distro: ", err)
+	for _, test := range tests {
+		tt := test
+		t.Run(test.name, func(t *testing.T) {
+			store, d, matcher := tt.setup()
+			actual, err := matcher.Match(store, &d, tt.p)
+			if tt.wantErr {
+				assert.Equal(t, "", err) //TODO: error case
```

Comment:
nit: it looks like the `wantErr` could probably be removed for now

Reply:
removed  👍